src/java/htsjdk/samtools/filter/AbstractJavascriptFilter.java

```diff
@@ -0,0 +1,173 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package htsjdk.samtools.filter;
+
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.io.Reader;
+
+import javax.script.Bindings;
+import javax.script.Compilable;
+import javax.script.CompiledScript;
+import javax.script.ScriptEngine;
+import javax.script.ScriptEngineManager;
+import javax.script.ScriptException;
+import javax.script.SimpleBindings;
+
+import htsjdk.samtools.util.CloserUtil;
+import htsjdk.samtools.util.IOUtil;
+import htsjdk.samtools.util.RuntimeIOException;
+
+/**
+ * Javascript filter with HEADER type containing TYPE records. contains two
+ * static method to get a SAM Read filter or a VariantFilter.
+ * 
+ * warning: tools, like galaxy, using this class are not safe because a script
+ * can access the filesystem.
+ * 
+ * @author Pierre Lindenbaum PhD
+ */
+public abstract class AbstractJavascriptFilter<HEADER, TYPE> {
+    public static final String DEFAULT_HEADER_KEY = "header";
+    /** compiled user script */
+    private CompiledScript script = null;
+
+    /** javascript bindings */
+    protected Bindings bindings;
+
+    /**
+     * constructor using a java.io.File script, compiles the script, puts
+     * 'header' in the bindings
+     */
+    protected AbstractJavascriptFilter(final File scriptFile, final HEADER header) {
+        this(header, scriptFile);
+    }
+
+    /**
+     * constructor using a java.lang.String script, compiles the script, puts
+     * 'header' in the bindings
+     */
+    protected AbstractJavascriptFilter(final String scriptExpression, final HEADER header) {
+        this(header, scriptExpression);
+    }
+
+    /**
+     * Constructor, compiles script, put header in the bindings
+     * 
+     * @param scriptFileOrObject
+     *            javascript source should be either a File or a String
+     * @param header
+     *            the header to be injected in the javascript context
+     */
+    private AbstractJavascriptFilter(final HEADER header, final Object scriptFileOrString) {
```

Comment:
this is the private constructor, it's called by the two constructors (String / File) .

I kept both constructors (File and String) this way because the latest JS engine compiler (nashorn in jdk8) can use the java.io.File information to resolve a relative path ( https://wiki.openjdk.java.net/display/Nashorn/Nashorn+extensions )

> Current executing script file, line, directory of the script can be accessed by these global (read-only) properties. This is useful in debug/tracing scripts. __DIR__ can be used like Unix 'dirname' command. Script can load script-file relative path resources using __DIR__.

If we use a java.io.Reader we will lose some information. 

What do you think ?